scripts/apps/dashboard/user-activity/components/UserActivityWidget.tsx

```diff
@@ -0,0 +1,411 @@
+            $rootScope: ng.get('$rootScope'),
+            $timeout: ng.get('$timeout'),
+            activityService: ng.get('activityService'),
+            api: ng.get('api'),
+            authoringWorkspace: ng.get(
+                'authoringWorkspace',
+            ) as AuthoringWorkspaceService,
+            cards: ng.get('cards'),
+            datetime: ng.get('datetime'),
+            desks: ng.get('desks'),
+            metadata: ng.get('metadata'),
+            search: ng.get('search'),
+            superdesk: ng.get('superdesk'),
+        };
+
+        this.state = {
+            userSearchField: '',
+            searchOpen: true,
+            groups: null,
+            groupsData: null,
+            loading: false,
+        };
+    }
+
+    componentDidMount() {
+        [
+            'item:lock',
+            'item:unlock',
+            'item:spike',
+            'item:move',
+            'item:publish',
+            'item:update',
+        ].map((event) =>
+            this.services.$rootScope.$on(event, this.refreshItems.bind(this)),
+        );
+    }
+
+    refreshItems() {
+        if (this.state.user) {
+            this.fetchGroupsData();
+        }
+    }
+
+    async fetchItems(group: IGroup) {
+        const {api, search} = this.services;
+        const {_items} = typeof group.query === 'function'
+            ? await group.query((repo, filters) => genericFetch({search, api}, repo, filters))
+            : await genericFetch({search, api}, group.repo, group.query);
+
+        const itemIds = [];
+        const itemsById = {};
+
+        for (const item of _items) {
+            itemIds.push(item._id);
+            itemsById[item._id] = item;
+        }
+
+        return {
+            id: group.id,
+            itemIds,
+            itemsById,
+        };
+    }
+
+    fetchGroupsData() {
+        const {groups} = this.state;
+        const promises = groups.map((group) => this.fetchItems(group));
+
+        Promise.all(promises).then((data) => {
+            this.setState({groupsData: data, loading: false});
+        });
+    }
+
+    renderGroup(group: IGroup) {
+        const {groups, groupsData} = this.state;
+        const data = groupsData.find((g) => g.id === group.id);
+
+        if (!data) {
+            console.warn(
+                `Tried to render group '${group.id}' but no data was found`,
+            );
+            return null;
+        }
+
+        return (
+            <>
+                <div className="stage">
+                    <div className="stage-header">
+                        <button
+                            className={`stage-header__toggle ${
+                                group.collapsed ? 'closed' : ''
+                            }`}
+                            onClick={() => {
+                                const newGroups = groups.map((g) => {
+                                    if (g.id === group.id) {
+                                        g.collapsed = !g.collapsed;
```

Comment:
don't mutate React state

```return {...g, collapsed: !g.collapsed}```

Reply:
oops! It wasn't part of the state before, but now you're right